### DISCUSSION TITLE
Load times

### DISCUSSION THREAD

**Discussion by ed2050** [zauberzeug/nicegui#3533] [Q&A]
### Question

Are there any good ways to improve import time of nicegui module?  Does [MATPLOTLIB](https://nicegui.io/documentation/section_configuration_deployment#environment_variables) work for anyone?  Setting it makes no difference in my tests.

My import time is consistently 1.3 or 2.0 seconds (likely due to caching) regardless of whether MATPLOTLIB is set.  The correct code path in `elements/pyplot.py` is being run, but load time doesn't change.

**Comment by falkoschindler** [comment]
Hi @ed2050,

You could try uninstalling Matplotlib to check if the `MATPLOTLIB` flag works. Setting it to `false` should allow running NiceGUI, while omitting it should crash because Matplotlib can't be imported.

Apart from that I'm not sure how to improve load times. We'd need to find out which libraries are expensive to import and what to do about it. Maybe we can add more flags like `MATPLOTLIB` or postpone the import calls.

**Reply by ed2050** [reply]
Thanks @falkoschindler.  Your reply got me thinking...

[Lazy loading](https://stackoverflow.com/a/78312617/7663296) is one solution for dealing with expensive imports.  It can be a great help for optional imports that often aren't used, like matplotlib (however see data below - matplotlib has a very tiny import time on my machine).  And no need for user flags: if you never use it then you don't pay the import tax.  Win-win.

Lazy loading can also help with startup times for required modules, by shifting import from startup to first use.  But that could hurt responsiveness, so needs careful thought.

The best candidates for improvement seem to be markdown2, aggrid, nicegui.upload, and nicegui.code, based on data below.  However it's uncertain how much difference these would make.

# Data 

Let's start with data.  Here are the top 11 longest module imports (first value, in microseconds) according to `python3 -X importtime -c 'import nicegui'`.  Every other import is under 10 ms.

```
TOTAL import time: 1.967 sec
self [us] | cumulative | imported package
   376866 |     520167 |                   fastapi.openapi.models
    55089 |      56630 |         markdown2
    51825 |      55726 |               pyarrow.lib
    34782 |      34782 |                     jinja2.utils
    30260 |      44646 |                 pyarrow.compute
    29036 |      29036 |                   numpy.core._multiarray_umath
    28111 |     107535 |                       fastapi.exceptions
    17382 |      25731 |                 bidict
    11515 |     309899 |       nicegui.events
    10505 |      10505 |                             pydantic_core.core_schema
    10193 |      10193 |                                   urllib3.util.url
```

# Analysis

So fastapi models is biggest contributor.  It takes 376 ms for 20% of startup time.  Don't think you can do anything about that.  Nicegui code never references fastapi.openapi.models.  It seems to be a fastapi internal import.

No other module contributes more than 3% of startup time.  markdown2 may be a good candidate for lazy loading, but it will only save 50 ms at best.

FYI matplot contributes next to nothing on my machine, less than 1 ms:
```
      186 |        186 |           matplotlib
       34 |        220 |         matplotlib.pyplot
```

No wonder the MATPLOTLIB flag makes no difference for me.  Is it expensive for others?

# Cumulative time

Cumulative import numbers look a bit different.  Top scores, nicegui packages only (second number):
```
self [us] | cumulative | imported package
      908 |    1894718 | nicegui
     2657 |    1893478 |   nicegui.ui
     1536 |    1146812 |     nicegui.element
      433 |     777137 |       nicegui.json
      561 |     750788 |         nicegui.json.orjson_wrapper
      743 |     390300 |     nicegui.elements.aggrid
    11515 |     309899 |       nicegui.events
      325 |     297647 |         nicegui.background_tasks
      640 |     297323 |           nicegui.globals
      748 |     126522 |     nicegui.run
      519 |     125240 |       nicegui.air
      669 |     102268 |     nicegui.elements.upload
     4133 |     101600 |       nicegui.nicegui
      418 |      69607 |     nicegui.elements.code
```

If you could lazy load a few of these, it may help.  But also, it may not.  Many of these scores are likely inflated by first import of a shared module.  I.e. nicegui.json is so high (777 ms) because it imports fastapi (639 ms).  If you eliminate nicegui.json, then another nicegui module will still import fastapi and shift the import time there.  

# Conclusion

There's no clear and simple way to improve nicegui import time.  `import nicegui` triggers 1600 module imports, which is typical for a package of this size & complexity.  Most of those imports contribute a very small amount to the total (1250 are under 1 ms).  Biggest contributor is fastapi internal modules, which nicegui can't do anything about.

If one were to improve load times, the best candidates seem to be optional modules that are biggest contributors.  I would look at markdown2, aggrid, nicegui.upload, and nicegui.code.  Most nicegui projects probably don't use those modules (assuming they aren't called internally by other nicegui code).

**Reply by falkoschindler** [reply]
Thanks for this thorough analysis, @ed2050!
I don't remember 100%, but I think on some systems Matplotlib is tricky and costly to install. So disabling it via `MATPLOTLIB` is not only to reduce import times, but also to improve compatibility with certain architectures.
Anyway, it's good to see that there's no urgent need to improve the import mechanism.

**Comment by ed2050** [comment]
I tested commenting out several modules with longest imports in nicegui/ui.py.  Only aggrid makes a significant difference (best of 3):
- none / baseline - 1.32, 1.31, 1.32 sec
- aggrid - 1.03, 0.99, 1.01 sec

Saving 300 ms is significant and noticeable.  Suggest making aggrid lazy load to improve startup times.

**Reply by falkoschindler** [reply]
Strange. When I comment out aggrid in `ui.py`, I don't notice any difference. Its "self" time is only around 300 us.
And looking into aggrid.py, I don't see why importing this file should take so long.

**Reply by ed2050** [reply]
Self time is small but cumulative is large.  aggrid must pull in expensive modules that aren't imported elsewhere.

The culprit is pandas.  In my version, pandas is only imported by aggrid.py.  Hence why commenting out aggrid speeds up on my machine.

```
self [us] | cumulative | imported package
      743 |     390300 |     nicegui.elements.aggrid
      984 |     388609 |       pandas
```

In current repo, pandas is also imported by table.  If you comment out aggrid and table in ui.py, then you should see the speedup.

# Solution
aggrid and table use `import pandas` to test for existence of pandas module.  Lazy loading won't help with that.  Here's a much faster way to [check whether a module exists](https://stackoverflow.com/a/14050282).

Suggest changing aggrid.py and table.py as follows:

```py
#try:
#    import pandas as pd
#    globals.optional_features.add('pandas')
#except ImportError:
#    pass

import importlib
# find_spec returns none if module doesn't exist
importlib.util.find_spec ('pandas') and globals.optional_features.add ('pandas')
```

If pandas module is actually used anywhere, then lazy load `pd`.  I only see it used in type hints, so you may not need to load it at all.

**Reply by falkoschindler** [reply]
Using `importlib.util.find_spec` is a great idea, @ed2050! I used it to create PR #3548. What do you think?

**Reply by ed2050** [reply]
Looks good to me.  Now if I could just find a way to make fastapi stop using pydantic.BaseModel... 😅

**Reply by ed2050** [reply]
@falkoschindler You can do same with numpy in json/orjson_wrapper.py and json/builtin_wrapper.py.  Saves another 109 ms on my machine.

**Reply by falkoschindler** [reply]
See PR #3559 ✔️

**Reply by ed2050** [reply]
I found one more.  Last one, after this returns diminish substantially.

httpx is only used in `air.py`.  I don't think air hosting is that common.  Moving `import httpx` from top level to `class Air.__init__` saves about ~100 ms load time on my machine, just like numpy.

**Reply by falkoschindler** [reply]
See PR #3562 ✔️

**Reply by rodja** [reply]
Maybe you can ask over at FastAPI if they can do their imports also "lazy"...

**Reply by ed2050** [reply]
haha definitely thought about it.  Unfortunately don't think it will help.

Biggest contributor is fastapi.openapi.models (380ms or 61%).  Less than 1ms of that is imports.  The rest is deriving classes from pydantic.BaseModel, which calls tons of `pydantic/_internal` functions taking 20ms or less.  Combined with things like 88k calls to dict.copy (23 ms) and 234k calls to dict.get (22 ms).  No big bottlenecks, just grinding out 5ms improvements in many places.

Pydantic is a real mess.  If I wanted a typed language, I wouldn't be using python.
